Monacs.Core/Either.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Monacs.Core
+{
+    public interface IEither
+    {
+        LeftOrRight LeftOrRight { get; }
+        object Left { get; }
+        object Right { get; }
+    }
+
+    public readonly struct Either<L, R> : IEither, IEquatable<Either<L, R>>, IEquatable<IEither>
+    {
+        object IEither.Left => LeftValue;
+        object IEither.Right => RightValue;
+
+        public L LeftValue { get; }
+        public R RightValue { get; }
+        public LeftOrRight LeftOrRight { get; }
+        public bool IsLeft => LeftOrRight == LeftOrRight.Left;
+        public bool IsRight => LeftOrRight == LeftOrRight.Right;
+
+        private Either(L left, R right, LeftOrRight leftOrRight)
+        {
+            LeftValue = left;
+            RightValue = right;
+            LeftOrRight = leftOrRight;
+        }
+
+        public Either(L left) : this(left, default, LeftOrRight.Left) { }
+        public Either(R right) : this(default, right, LeftOrRight.Right) { }
+
+        public bool Equals(Either<L, R> other) =>
+            LeftOrRight == other.LeftOrRight
+            && IsLeft ? EqualityComparer<L>.Default.Equals(LeftValue, other.LeftValue)
+                      : EqualityComparer<R>.Default.Equals(RightValue, other.RightValue);
+
+        public bool Equals(IEither other) =>
+            !ReferenceEquals(null, other)
+            && LeftOrRight == other.LeftOrRight
+            && (other.Left != null && other.Left.Equals(LeftValue)
+                || other.Right != null && other.Right.Equals(RightValue));
+
+        public override bool Equals(object obj) =>
+            !ReferenceEquals(null, obj)
+            && obj is IEither either && Equals(either);
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = EqualityComparer<L>.Default.GetHashCode(LeftValue);
+                hashCode = (hashCode * 397) ^ EqualityComparer<R>.Default.GetHashCode(RightValue);
+                hashCode = (hashCode * 397) ^ (int)LeftOrRight;
+                return hashCode;
+            }
+        }
+
+        public static bool operator ==(in Either<L, R> a, in Either<L, R> b) => a.Equals(b);
+
+        public static bool operator !=(in Either<L, R> a, in Either<L, R> b) => !a.Equals(b);
+
+        public override string ToString()
+        {
+            var side = IsLeft ? "Left" : "Right";
+            var value = IsLeft ? LeftValue.ToString() : RightValue.ToString();
+            return $"Either<{typeof(L).Name}, {typeof(R).Name}>({side} = {value})";
+        }
+    }
+
+    public enum LeftOrRight { Left, Right }
+
+    public static partial class Either
```

Comment:
I would use the same convention for generic parameter names as in other places in this library, so I would recommend using `TLeft`, `TRight`, `TOut`, `TOutLeft`, `TOutRight` respectively.

I would also move the extensions to separate file (and remove partial as it's not needed here.

Reply:
Yes, I left this kind of things for the end :)